syntax/lexer.go

```diff
@@ -346,6 +346,29 @@ skipSpace:
 	}
 }
 
+// tokenizeGlob determines whether the expression should be tokenized as a glob literal
+func (p *Parser) tokenizeGlob() bool {
+	if p.val == "function" {
+		return false
+	}
+	// NOTE: empty pattern list is a valid globbing syntax, eg @()
+	// but we'll operate on the "likelihood" that it is a function;
+	// only tokenize if its a non-empty pattern list
+	if p.peekBytes("()") {
+		return false
+	}
+	return p.peekByte('(')
+}
+
+func (p *Parser) peekBytes(s string) bool {
+	if p.bsp == len(p.bs) {
```

Comment:
use len(s) here

Reply:
hmm sorry, I don't think I understand. my understanding is, with `p.bsp == len(p.bs)`, the intention is to load more bytes from the source if we're at the last bit. 

do you mean `len(s) > len(p.bs)` as in, if `p.bs` doesn't have enough bytes to compare with `s` load more to the buffer? wouldn't `p.bsp == len(p.bs) || len(s) > len(p.bs)` make more sense?

Reply:
Think of `peekByte(b byte)` as a shortcut for `peekBytes(s string)` where `len(s) == 1`. It does `if p.bsp == len(p.bs)` as an equivalent to `if p.bsp >= len(p.bs)`; it just needs one byte, and `p.bsp` is the position of the next byte within `bs`, so that's enough.

In `peekBytes`, you want to peek `len(s)` bytes, not just one. So the right check seems to be `if p.bsp + (len(bs)-1) >= len(p.bs)`.

Another question here is how `p.fill` only guarantees to fill at least one byte. So I think you should swap `if` for `for`; otherwise, if you need to peek 2 bytes and you have 0, one `fill` call could fill just one byte, so you'd need to call it again.

tl;dr:

```
for p.bsp + (len(bs)-1) >= len(p.bs) {
    p.fill()
}
```

Reply:
I totally misunderstood. Thanks for the thorough explanation :)